frontend/HelperFuncs/personalization.js

```diff
@@ -21,15 +19,268 @@ preconditions:
+tests are either left or right tailed depending on the content of [category] to capture the different types of goals
+*/
+function isStatisticallySignificant(userCategoryPercent, similarCategoryData, category){
+    const population = [...similarCategoryData, userCategoryPercent]
+    const populationStdDev = findPopulationStandardDeviation(userCategoryPercent, similarCategoryData);
+    // standard deviation cannot be negative
+    if(populationStdDev < 0){
+        throw new Error('negative standard deviation')
+    }
+    const populationMean = getAverage(population)
+    const zScore = (userCategoryPercent - populationMean)/populationStdDev
+    /* if goals whose category is about reducing spending, zScore should be on the positive tail
+    i.e. above average to be relevant and negative for increase spending categories */
+    if(zScore < 0 && (category !== 'investment' || category !== 'savings_account' || category !== 'financial_planning')){
+        return false
+    }
+    if(zScore > 0 && (category === 'investment' || category === 'savings_account' || category === 'financial_planning')){
+        return false
+    }
+    const pValue = standardNormalPDF(zScore, populationStdDev, populationMean)
+    // pValues must always be in [0,1] if they're not, something's gone wrong
+    if (pValue > 1 || pValue < 0){
+        throw new Error('p value out of range')
+    }
+    // a .05 alpha gives us 95% confidence in our results
+    const alpha = 0.05
+    return pValue < alpha
+
+}
+
+
+/*
+this function aggregates all similarUsers in the list [similarUsers] expenditure percentages for subcaegory [category]. 
+this is done by implementing an iterative variant of depth-first search to find similar user category data
+inside the nested details or "subcategory" objects. Returns a list of floats. 
+*/
+function findSimilarUserCategoryData(similarUsers, category){
+    const matchingSimilarData = []
+    similarUsers.forEach(similarTransaction => {
+        const transaction = similarTransaction.transaction
+        for(let parentCategory in transaction){
+            const details = transaction[parentCategory].details
+            for (const childCategory of details) { 
+                if(childCategory.name === category){
+                    matchingSimilarData.push(childCategory.percent)
+                    }
+            }
+        }
+    });
+    return matchingSimilarData;
+}
+
 
+/*
 
-    return []
+*/
+function findUserCategoryData(userData, category) {
+    for (let parentCategory in userData) {
+        const details = userData[parentCategory].details;
+        for (const childCategory of details) {
+            if (childCategory.name === category) {
+                return childCategory.percent;
+            }
+        }
+    }
 }
 
 
+/*
+returns the mean value of [list] where [list] is an array of numbers 
+*/
+function getAverage(list){
```

Comment:
I'm surprised the Math module doesn't already have an average function. nonetheless, good helper to reduce duplicate code.

Reply:
yeah I think javascipts built in math module is actually surprisingly lacking, probably because js isn't really a language you're going to be doing math in very often but I think that played out well here since it made this file implementation more involved